### PR TITLE
Update lsquic binding lifecycle

### PR DESCRIPTION
**Pull request by tenpura-shrimp**
* Implement engine_process_connections timer

The previous code did not have an implementation to set timers for processing
connections. This leads to an occasional deadlock where process_connections
must be called to continue. This change implements this timer modelled after
the lsquic reference implementation.

Reference: https://github.com/litespeedtech/lsquic/blob/master/bin/prog.c#L535

* Update recieve function to read full data from socket at once 

The existing code reads each message from the socket individually and calls
engine_process_conns after each message. This seems inefficient as it leads
to excess calls engine_process_conns.

This changes the code that reads from the socket to read all available messages
before calling engine_process_conns based off of the logic in the reference
implementation.

reference: https://github.com/litespeedtech/lsquic/blob/master/bin/test_common.c#L737

### CODE REVIEW

**Comment by tenpura-shrimp** [comment]
results from a test I ran of livestream manifest, measured from the browser

Latencies with the original code: Mean: 74ms
lowlatencyorig  fulllatencyoriginal

Latencies with the updated code: Mean: 67ms
lowlatencynew fullatencynew

File: src/lsquic/client.cr [hunk 1]
```diff
@@ -209,21 +211,55 @@ module QUIC
...
       end
 
+      @process_fiber = spawn do
+        loop do
+          sleep
+          LibLsquic.engine_process_conns(engine)
+          diff = 0
+          # check advisory time
+          if LibLsquic.engine_earliest_adv_tick(engine, pointerof(diff)) != 0
+            Thread.current.scheduler.@current.resume_event.add(diff.microseconds)
```
Comment (saltycrys) [inline src/lsquic/client.cr:229]:
```suggestion
            Crystal::Scheduler.current_fiber.resume_event.add(diff.microseconds)
```

**Review by saltycrys — APPROVED** [review]
This is really great!

**Comment by unixfox** [comment]
Does this need testing now then?

**Comment by Perflyst** [comment]
Yes, would be good to test it first.

**Comment by Perflyst** [comment]
Patch is now in testing on my server

**Comment by Perflyst** [comment]
As far as I can see @tenpura-shrimp force pushed the two commits after @saltycrys reviewed. Could you take a look again @saltycrys? Afterwards we can merge and release 2.23.1

**Review by saltycrys — APPROVED** [review]

**Pull request merged**